samplers/parser.go

```diff
@@ -346,120 +346,160 @@ func (p *Parser) ParseMetricSSF(metric *ssf.SSFSample) (UDPMetric, error) {
 
 // ParseMetric converts the incoming packet from Datadog DogStatsD
 // Datagram format in to a Metric. http://docs.datadoghq.com/guides/dogstatsd/#datagram-format
-func (p *Parser) ParseMetric(packet []byte) (*UDPMetric, error) {
-	ret := &UDPMetric{
+func (p *Parser) ParseMetric(packet []byte, cb func(*UDPMetric)) error {
+	metric := &UDPMetric{
 		SampleRate: 1.0,
 	}
-	pipeSplitter := NewSplitBytes(packet, '|')
-	pipeSplitter.Next() // first split always succeeds, since there are at least zero pipes
+	typeStart := bytes.IndexByte(packet, '|')
+	if typeStart < 0 {
+		return errors.New("Invalid metric packet, need at least 1 pipe for type")
+	}
 
-	startingColon := bytes.IndexByte(pipeSplitter.Chunk(), ':')
-	if startingColon == -1 {
-		return nil, errors.New("Invalid metric packet, need at least 1 colon")
+	valueStart := bytes.IndexByte(packet[:typeStart], ':')
+	if valueStart == -1 {
+		return errors.New("Invalid metric packet, need at least 1 colon")
 	}
-	nameChunk := pipeSplitter.Chunk()[:startingColon]
-	valueChunk := pipeSplitter.Chunk()[startingColon+1:]
+	nameChunk := packet[:valueStart]
+	valueChunk := packet[valueStart+1 : typeStart]
+
 	if len(nameChunk) == 0 {
-		return nil, errors.New("Invalid metric packet, name cannot be empty")
+		return errors.New("Invalid metric packet, name cannot be empty")
 	}
 
-	if !pipeSplitter.Next() {
-		return nil, errors.New("Invalid metric packet, need at least 1 pipe for type")
+	metric.Name = string(nameChunk)
+
+	tagsStart := len(packet)
+	if idx := bytes.IndexByte(packet[typeStart+1:], '|'); idx > -1 {
+		tagsStart = typeStart + 1 + idx
 	}
-	typeChunk := pipeSplitter.Chunk()
+	typeChunk := packet[typeStart+1 : tagsStart]
```

Comment:
again `bytes.Cut` might be cleaner than `bytes.IndexByte` for this logic (but maybe not! if you don't think so please ignore)